parser.py

```diff
@@ -159,39 +181,51 @@ def expressions(self):
 
     def expression(self):
         token = self.peek()
-        #print(token.token_type, token.value)
+
         exp = Expr()
         if isinstance(token, RealLiteral):  # if just a real number, return it
-            exp = RealConstant(token.value)
+            exp = RealConstant(float(token.value))
             self.remove()  # real
         if isinstance(token, IntLiteral):  # if just a integer, return it
-            exp = IntegerConstant(token.value)
+            exp = IntegerConstant(int(token.value))
             self.remove()  # integer
 
         if isinstance(token, Identifier):  # parameter/data
-            exp = Placeholder(token.value, None)
+            if token.value in self.data_names:
+                exp = Data(token.value)
+            else:
+                exp = Param(token.value)
             self.remove()  # identifier
 
         if PrefixOps.check(token):  # prefixOp expression
-            self.remove()  # op
+            self.expect_token(Operator, PrefixOps.ops)  # operator
+            self.remove()
+
             next_expression = self.expression()
             exp = PrefixOps.generate(next_expression, token)
 
         if isinstance(token, Special) and token.value == "(":  # '(' expression ')'
             self.remove()  # (
             next_expression = self.expression()
+            self.expect_token(Special, ")")  # )
             self.remove()  # )
             exp = next_expression  # expression
 
         next_token = self.peek()  # this is for the following 2 rules, which have conditions after expression
         if isinstance(next_token, Special) and next_token.value == "[":  # identifier '[' expressions ']'
-            self.remove()  # ]
+            self.remove()  # [
+            warnings.warn("Parser: Indices are assumed to be a single literal, not expression.")
             expressions = self.expressions()  # list of expression
+            self.expect_token(Special, "]")
             self.remove()  # ]
-            exp.index = Index(tuple(expressions))
+
+            # Assume index is a single identifier - this is NOT GOOD
+
+            exp.index = Index(tuple(expression.name for expression in  expressions))
```

Comment:
Now looking at the compiler's code it seems like indexes atm can only be a single literal. Do you think it would be possible to support full expressions? It probably is overkill for a simple demo but if the goal is to support a full program with complete grammar then it should be considered.

Reply:
I merged this PR cause it worked and it looks sick. We're going all the way from text to a probably very buggy model which is very exciting.

Two immediate issues are:
*. Figure out how lag operators should work
*. Figure out how parameter constraints should work